### PR TITLE
Ensure mbus is not shut down while processing mbus requests

### PR DESCRIPTION
**Pull request by jonmv**
@bjorncs please review. @vekterli @baldersheim FYI.

### CODE REVIEW

**Comment by jonmv** [comment]
The idea here is that a `Request` that is created by an `MbusServer` while its owning `Container` is about to be switched out, will instead reference the newly created `Container`—this potentially leads to destruction of the `MessageBus` that is used by the `MbusServer`, e.g., when sessions change as a result of a document type being removed. As a first step when handling a `Message`, `MessageBus` will push itself as a `ReplyHandler`; when handling the reply that is eventually returned from the `MbusServer`, though handling in the new `Container`, it may have been `destroyed()` together with its owning `Container`, and then the `Reply` will simply be `discard()`ed, leaving the client on the other side of the RPC connection to wait for ever.

**Comment by baldersheim** [comment]
Is this observed to be a problem currently => will this solve a real problem.
What is the risk that it will create a new set of problems ?

**Comment by jonmv** [comment]
We have an unstable system test that feeds while removing a document type. Messages sometimes go missing in the container precisely when it reconfigures. I believe this may be the cause. There's also another race when we switch containers: we should take care to not release the creation reference to the previous container before we have stopped its servers, synced with them to ensure any in-progress `Request` creation is complete, and activated the new servers. We don't do that, but that is the next commit.

File: jdisc_core/src/main/java/com/yahoo/jdisc/core/ApplicationLoader.java [hunk 1]
```diff
@@ -186,7 +186,7 @@ public void stop() throws Exception {
         synchronized (appLock) {
             application = null;
         }
-        activateContainer(null);
+        try (DeactivatedContainer deactivated = activateContainer(null)) { }
```
Comment (bjorncs) [inline jdisc_core/src/main/java/com/yahoo/jdisc/core/ApplicationLoader.java:189]: Can be simplified to `activacteContainer(null).close()`

File: jdisc_core/src/main/java/com/yahoo/jdisc/test/TestDriver.java [hunk 2]
```diff
@@ -73,9 +73,10 @@ public ContainerBuilder newContainerBuilder() {
         return loader.newContainerBuilder();
     }
 
+    /** Returns the deactivated container, with its container reference already released. */
     @Override
     public DeactivatedContainer activateContainer(ContainerBuilder builder) {
-        return loader.activateContainer(builder);
+        try (DeactivatedContainer deactivated = loader.activateContainer(buionf
```
Comment (bjorncs) [inline jdisc_core/src/main/java/com/yahoo/jdisc/test/TestDriver.java:79]: Typo

**Review by bjorncs — CHANGES_REQUESTED** [review]

File: jdisc_core/src/main/java/com/yahoo/jdisc/core/ApplicationLoader.java (hunk 1 above)
Comment (jonmv) [reply jdisc_core/src/main/java/com/yahoo/jdisc/core/ApplicationLoader.java:189]: `activateContainer(...)` returns `null` when there is no previous container.

File: jdisc_core/src/main/java/com/yahoo/jdisc/test/TestDriver.java (hunk 2 above)
Comment (jonmv) [reply jdisc_core/src/main/java/com/yahoo/jdisc/test/TestDriver.java:79]: What!?

**Review by bjorncs — APPROVED** [review]

File: jdisc_core/src/main/java/com/yahoo/jdisc/test/TestDriver.java (hunk 2 above)
Comment (jonmv) [reply jdisc_core/src/main/java/com/yahoo/jdisc/test/TestDriver.java]: Thanks.

File: jdisc_core/src/main/java/com/yahoo/jdisc/core/ApplicationLoader.java (hunk 1 above)
Comment (bjorncs) [reply jdisc_core/src/main/java/com/yahoo/jdisc/core/ApplicationLoader.java:189]: Clever!

**Pull request merged**